File: src/rhs/navierstokes.jl
"""
    NavierStokes{T}

Incompressible Navier–Stokes in spectral space, rotational form:

    dû/dt = P(k) [ (u × ω)^ + f̂ ] − ν k² û

Evaluation pipeline (all in-place, dominated by the FFTs):
  1. ω̂ = ik × û in the spectral scratch; transform ω̂, then a copy of û, to
     physical space (FFTW's multidim c2r transforms destroy their input even
     out-of-place, so û itself is never transformed — this ordering is
     load-bearing)
  2. compute u × ω pointwise in physical space (own buffer `nl_phys`)
  3. forward-transform into `dû` and add the forcing — BEFORE projection, so
     a non-solenoidal forcing's gradient part is absorbed into the pressure
     (the ik·f̂/|k|² term of the eliminated pressure equation)
  4. dealias (2/3 mask, which also zeroes k = 0) and project divergence-free
  5. add the viscous term −ν k² û (û stays exactly dealiased, so it needs no
     extra masking)

Owns its scratch buffers so multiple RHS objects can share one `Grid`.
"""
struct NavierStokes{T, F <: AbstractForcing, G <: Grid, AP, AS} <: AbstractRHS
    ν::T              # kinematic viscosity
    forcing::F
    grid::G
    # --- scratch buffers (preallocated on the grid's backend) ---
    u_phys::AP        # physical-space velocity    (Nx, Ny, Nz, 3)
    ω_phys::AP        # physical-space vorticity   (Nx, Ny, Nz, 3)
    nl_phys::AP       # physical-space u × ω       (Nx, Ny, Nz, 3)
    scratch_spec::AS  # spectral work array        (Nx÷2+1, Ny, Nz, 3)
end

"""
    NavierStokes(g::Grid; ν, forcing = NoForcing())

Construct the RHS and allocate its scratch buffers on the grid's backend.
"""
function NavierStokes(g::Grid; ν, forcing = NoForcing())
    T = eltype(g.k2)
    u_phys  = similar(g.k2, T, (g.Nx, g.Ny, g.Nz, 3))
    ω_phys  = similar(u_phys)
    nl_phys = similar(u_phys)
    scratch_spec = similar(g.k2, Complex{T}, (g.Nx ÷ 2 + 1, g.Ny, g.Nz, 3))
    return NavierStokes(T(ν), forcing, g, u_phys, ω_phys, nl_phys,
                        scratch_spec)
end

function rhs!(dû, û, r::NavierStokes, t)
    g = r.grid
    ŝ = r.scratch_spec

    û1 = view(û, :, :, :, 1); û2 = view(û, :, :, :, 2); û3 = view(û, :, :, :, 3)
    ŝ1 = view(ŝ, :, :, :, 1); ŝ2 = view(ŝ, :, :, :, 2); ŝ3 = view(ŝ, :, :, :, 3)

    # 1. ω̂ = ik × û  (into the spectral scratch)
    @. ŝ1 = im * (g.ky * û3 - g.kz * û2)
    @. ŝ2 = im * (g.kz * û1 - g.kx * û3)
    @. ŝ3 = im * (g.kx * û2 - g.ky * û1)

    # 2. to physical space. c2r destroys its input: ω̂ first (expendable),
    #    then a copy of û — never û itself.
    mul!(r.ω_phys, g.iplan, ŝ)
    copyto!(ŝ, û)
    mul!(r.u_phys, g.iplan, ŝ)

    # 3. nonlinear term u × ω, pointwise
    u1 = view(r.u_phys, :, :, :, 1); ω1 = view(r.ω_phys, :, :, :, 1)
    u2 = view(r.u_phys, :, :, :, 2); ω2 = view(r.ω_phys, :, :, :, 2)
    u3 = view(r.u_phys, :, :, :, 3); ω3 = view(r.ω_phys, :, :, :, 3)
    n1 = view(r.nl_phys, :, :, :, 1)
    n2 = view(r.nl_phys, :, :, :, 2)
    n3 = view(r.nl_phys, :, :, :, 3)
    @. n1 = u2 * ω3 - u3 * ω2
    @. n2 = u3 * ω1 - u1 * ω3
    @. n3 = u1 * ω2 - u2 * ω1

    # 4. back to spectral space; forcing enters before dealias/projection
    mul!(dû, g.plan, r.nl_phys)
    apply_forcing!(dû, û, r.u_phys, r.forcing, g, t)

    # 5. dealias (also zeroes k = 0) and project; ŝ is free again after the
    #    transforms, so its first component serves as the projection scratch
    dealias!(dû, g)
    project!(dû, g, ŝ1)

    # 6. viscous term
    @. dû -= r.ν * g.k2 * û
    return dû
end

"""Diagonal viscous operator `-ν k²` for integrating-factor schemes.
Allocates a fresh array on each call."""
linear_operator(r::NavierStokes) = -r.ν .* r.grid.k2

File: src/hit3d/navierstokes.py
"""Incompressible Navier–Stokes in spectral space, rotational form:

    dû/dt = P(k) [ (u × ω)^ + f̂ ] − ν k² û

`make_ns_rhs(grid)` returns a pure function ``rhs(u_hat, params, t)``;
`NSParams` carries the differentiable physical scalars (ν and the forcing's
coefficients), so gradients w.r.t. them need no refactoring — just
``jax.grad`` with the params pytree as the argument.

Evaluation pipeline (XLA owns all buffers and FFT plans):
  1. ω̂ = ik × û; transform ω̂ and û to physical space
  2. compute u × ω pointwise in physical space
  3. forward-transform and add the forcing — BEFORE projection, so a
     non-solenoidal forcing's gradient part is absorbed into the pressure
     (the ik·f̂/|k|² term of the eliminated pressure equation)
  4. dealias (2/3 mask, which also zeroes k = 0) and project
     divergence-free
  5. add the viscous term −ν k² û (û stays exactly dealiased, so it needs
     no extra masking)

Call `make_ns_rhs` once per grid and reuse the returned function: jitted
callers (`rollout`, `evolve`) treat it as a static argument, so every new
closure triggers a fresh compilation.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from functools import partial

import jax
import jax.numpy as jnp

from .forcing import NoForcing
from .grid import dealias, project, to_physical, to_spectral

__all__ = ["NSParams", "make_ns_rhs", "linear_operator"]


@partial(
    jax.tree_util.register_dataclass,
    data_fields=["nu", "forcing"],
    meta_fields=[],
)
@dataclass(frozen=True)
class NSParams:
    """Differentiable parameters of the Navier–Stokes RHS: the kinematic
    viscosity `nu` plus the forcing (whose scalar coefficients are pytree
    leaves of their own)."""

    nu: jax.Array
    forcing: object = field(default_factory=NoForcing)

    @property
    def label(self):
        return f"NavierStokes_nu{float(self.nu):g}_{self.forcing.label}"

    def __repr__(self):
        return f"NavierStokes(nu={float(self.nu):g}, forcing={self.forcing!r})"


def make_ns_rhs(grid):
    """Build the rotational-form Navier–Stokes RHS on `grid`.

    Returns ``rhs(u_hat, params: NSParams, t) -> du_hat``, pure and
    jit/grad-transparent.
    """

    def rhs(u_hat, params, t):
        kx, ky, kz = grid.kx, grid.ky, grid.kz

        # 1. ω̂ = ik × û, then both to physical space
        w_hat = 1j * jnp.stack((
            ky * u_hat[2] - kz * u_hat[1],
            kz * u_hat[0] - kx * u_hat[2],
            kx * u_hat[1] - ky * u_hat[0],
        ))
        u = to_physical(u_hat, grid)
        w = to_physical(w_hat, grid)

        # 2. nonlinear term u × ω, pointwise
        nl = jnp.stack((
            u[1] * w[2] - u[2] * w[1],
            u[2] * w[0] - u[0] * w[2],
            u[0] * w[1] - u[1] * w[0],
        ))

        # 3. back to spectral space; forcing enters before dealias/projection
        du = to_spectral(nl, grid) + params.forcing.term(u_hat, grid, t)

        # 4. dealias (also zeroes k = 0) and project divergence-free
        du = project(dealias(du, grid), grid)

        # 5. viscous term
        return du - params.nu * grid.k2 * u_hat

    return rhs


def linear_operator(grid, params):
    """Diagonal viscous operator −ν k² for future integrating-factor / ETD
    schemes."""
    return -params.nu * grid.k2
